engine/class_modules/priest/sc_priest_shadow.cpp

```diff
@@ -2227,34 +2221,55 @@ namespace buffs
 // ==========================================================================
 struct voidform_t final : public priest_buff_t<buff_t>
 {
-  voidform_t( priest_t& p ) : base_t( p, "voidform", p.specs.voidform )
+  double am_value;
+
+  voidform_t( priest_t& p )
+    : base_t( p, "voidform", p.specs.voidform ),
+      am_value( priest().is_ptr() ? priest().talents.shadow.ancient_madness->effectN( 2 ).percent() / 10
+                                  : priest().talents.shadow.ancient_madness->effectN( 2 ).percent() )
   {
     add_invalidate( CACHE_PLAYER_DAMAGE_MULTIPLIER );
     add_invalidate( CACHE_PLAYER_HEAL_MULTIPLIER );
 
-    // Set cooldown to 0s, cooldown is stored in Void Eruption
-    cooldown->duration = timespan_t::from_seconds( 0 );
+    if ( priest().talents.shadow.ancient_madness.enabled() )
+    {
+      add_invalidate( CACHE_CRIT_CHANCE );
+      add_invalidate( CACHE_SPELL_CRIT_CHANCE );
+      set_reverse( true );
+      set_period( data().effectN( 4 ).period() );
+      set_max_stack( 20 );
+      set_default_value( am_value );
 
-    // Use a stack change callback to trigger voidform effects.
-    set_stack_change_callback( [ this ]( buff_t*, int, int cur ) {
-      if ( cur )
-      {
-        priest().cooldowns.mind_blast->reset( true, -1 );
-        priest().cooldowns.void_bolt->reset( true );
-      }
-    } );
+      // When Ancient Madness expires, make sure we do not expire Voidform
+      set_stack_change_callback( [ this ]( buff_t* b, int old, int cur ) {
+        if ( cur == 1 )
+        {
+          set_reverse( false );
+          set_max_stack( 1 );
+          set_default_value( 0 );
+        }
+      } );
+    }
```

Comment:
This is a slight hack, without this when the buff tries to "remove" the crit portion it would also expire Voidform. We use this to say, actually stop giving the crit bonus and let the buff finish out.